File: historian/src/storage/codec.ts
import { Point, ValueTypeCode, fixedTypeSize } from "../types/valueTypes";

export const SEGMENT_HEADER_SIZE = 17; // u32 tagId + i64 ts + u8 type + u32 valueLen
export const BLOCK_INDEX_ENTRY_SIZE = 44;

export interface DecodedRecord {
  point: Point;
  bytesRead: number;
}

export interface BlockIndexEntry {
  minTs: bigint;
  maxTs: bigint;
  byteOffsetStart: bigint;
  byteOffsetEnd: bigint;
  pointCount: number;
  minTagId: number;
  maxTagId: number;
}

export function encodeUdpBatch(points: Point[]): Buffer {
  const chunks: Buffer[] = [];
  const head = Buffer.allocUnsafe(4);
  head.writeUInt32LE(points.length, 0);
  chunks.push(head);
  for (const p of points) {
    chunks.push(encodeUdpPoint(p));
  }
  return Buffer.concat(chunks);
}

export function encodeUdpPoint(point: Point): Buffer {
  const typeSize = fixedTypeSize(point.typeCode);
  const valueBuf =
    typeSize > 0
      ? encodeFixedValue(point.typeCode, point.value as number)
      : Buffer.from(point.value as string, "utf8");
  const extra = typeSize > 0 ? 0 : 4;
  const b = Buffer.allocUnsafe(4 + 8 + 1 + extra + valueBuf.length);
  let o = 0;
  b.writeUInt32LE(point.tagId >>> 0, o);
  o += 4;
  b.writeBigInt64LE(point.tsEpoch, o);
  o += 8;
  b.writeUInt8(point.typeCode, o);
  o += 1;
  if (typeSize < 0) {
    b.writeUInt32LE(valueBuf.length, o);
    o += 4;
  }
  valueBuf.copy(b, o);
  return b;
}

export function decodeUdpBatch(packet: Buffer): Point[] {
  if (packet.length < 4) throw new Error("packet too short");
  const count = packet.readUInt32LE(0);
  const out: Point[] = [];
  let o = 4;
  for (let i = 0; i < count; i++) {
    if (o + 13 > packet.length) throw new Error("incomplete point header");
    const tagId = packet.readUInt32LE(o);
    o += 4;
    const tsEpoch = packet.readBigInt64LE(o);
    o += 8;
    const typeCode = packet.readUInt8(o) as ValueTypeCode;
    o += 1;
    const size = fixedTypeSize(typeCode);
    let value: string | number;
    if (size > 0) {
      if (o + size > packet.length) throw new Error("incomplete point value");
      value = decodeFixedValue(packet, o, typeCode);
      o += size;
    } else {
      if (o + 4 > packet.length) throw new Error("incomplete string length");
      const len = packet.readUInt32LE(o);
      o += 4;
      if (o + len > packet.length) throw new Error("incomplete string payload");
      value = packet.toString("utf8", o, o + len);
      o += len;
    }
    out.push({ tagId, tsEpoch, typeCode, value });
  }
  return out;
}

export function encodeSegmentRecord(point: Point): Buffer {
  const typeSize = fixedTypeSize(point.typeCode);
  const valueBuf =
    typeSize > 0
      ? encodeFixedValue(point.typeCode, point.value as number)
      : Buffer.from(point.value as string, "utf8");
  const valueLen = typeSize > 0 ? 0 : valueBuf.length;
  const out = Buffer.allocUnsafe(SEGMENT_HEADER_SIZE + valueBuf.length);
  out.writeUInt32LE(point.tagId >>> 0, 0);
  out.writeBigInt64LE(point.tsEpoch, 4);
  out.writeUInt8(point.typeCode, 12);
  out.writeUInt32LE(valueLen >>> 0, 13);
  valueBuf.copy(out, SEGMENT_HEADER_SIZE);
  return out;
}

export function decodeSegmentRecord(buf: Buffer, offset: number): DecodedRecord | null {
  if (offset + SEGMENT_HEADER_SIZE > buf.length) return null;
  const tagId = buf.readUInt32LE(offset);
  const tsEpoch = buf.readBigInt64LE(offset + 4);
  const typeCode = buf.readUInt8(offset + 12) as ValueTypeCode;
  const valueLen = buf.readUInt32LE(offset + 13);
  const fixed = fixedTypeSize(typeCode);
  const payloadLen = fixed > 0 ? fixed : valueLen;
  const total = SEGMENT_HEADER_SIZE + payloadLen;
  if (offset + total > buf.length) return null;
  const payloadOffset = offset + SEGMENT_HEADER_SIZE;
  const value =
    fixed > 0 ? decodeFixedValue(buf, payloadOffset, typeCode) : buf.toString("utf8", payloadOffset, payloadOffset + payloadLen);
  return {
    point: { tagId, tsEpoch, typeCode, value },
    bytesRead: total
  };
}

export function encodeBlockIndexEntry(e: BlockIndexEntry): Buffer {
  const out = Buffer.allocUnsafe(BLOCK_INDEX_ENTRY_SIZE);
  out.writeBigInt64LE(e.minTs, 0);
  out.writeBigInt64LE(e.maxTs, 8);
  out.writeBigUInt64LE(e.byteOffsetStart, 16);
  out.writeBigUInt64LE(e.byteOffsetEnd, 24);
  out.writeUInt32LE(e.pointCount >>> 0, 32);
  out.writeUInt32LE(e.minTagId >>> 0, 36);
  out.writeUInt32LE(e.maxTagId >>> 0, 40);
  return out;
}

export function decodeBlockIndex(buf: Buffer): BlockIndexEntry[] {
  const n = Math.floor(buf.length / BLOCK_INDEX_ENTRY_SIZE);
  const out: BlockIndexEntry[] = [];
  for (let i = 0; i < n; i++) {
    const off = i * BLOCK_INDEX_ENTRY_SIZE;
    out.push({
      minTs: buf.readBigInt64LE(off),
      maxTs: buf.readBigInt64LE(off + 8),
      byteOffsetStart: buf.readBigUInt64LE(off + 16),
      byteOffsetEnd: buf.readBigUInt64LE(off + 24),
      pointCount: buf.readUInt32LE(off + 32),
      minTagId: buf.readUInt32LE(off + 36),
      maxTagId: buf.readUInt32LE(off + 40)
    });
  }
  return out;
}

function encodeFixedValue(typeCode: ValueTypeCode, value: number): Buffer {
  const s = fixedTypeSize(typeCode);
  const b = Buffer.allocUnsafe(s);
  switch (typeCode) {
    case ValueTypeCode.Int8:
      b.writeInt8(value, 0);
      return b;
    case ValueTypeCode.UInt8:
      b.writeUInt8(value, 0);
      return b;
    case ValueTypeCode.Int16:
      b.writeInt16LE(value, 0);
      return b;
    case ValueTypeCode.UInt16:
      b.writeUInt16LE(value, 0);
      return b;
    case ValueTypeCode.Int32:
      b.writeInt32LE(value, 0);
      return b;
    case ValueTypeCode.UInt32:
      b.writeUInt32LE(value, 0);
      return b;
    case ValueTypeCode.Float32:
      b.writeFloatLE(value, 0);
      return b;
    case ValueTypeCode.Float64:
      b.writeDoubleLE(value, 0);
      return b;
    default:
      throw new Error(`Type ${typeCode} is not fixed-size`);
  }
}

function decodeFixedValue(buf: Buffer, offset: number, typeCode: ValueTypeCode): number {
  switch (typeCode) {
    case ValueTypeCode.Int8:
      return buf.readInt8(offset);
    case ValueTypeCode.UInt8:
      return buf.readUInt8(offset);
    case ValueTypeCode.Int16:
      return buf.readInt16LE(offset);
    case ValueTypeCode.UInt16:
      return buf.readUInt16LE(offset);
    case ValueTypeCode.Int32:
      return buf.readInt32LE(offset);
    case ValueTypeCode.UInt32:
      return buf.readUInt32LE(offset);
    case ValueTypeCode.Float32:
      return buf.readFloatLE(offset);
    case ValueTypeCode.Float64:
      return buf.readDoubleLE(offset);
    default:
      throw new Error(`Type ${typeCode} is not fixed-size`);
  }
}

File: historian_v2/src/storage/codec.rs
use anyhow::{bail, Result};

use crate::types::{Point, PointValue, ValueTypeCode};

pub const SEGMENT_HEADER_SIZE: usize = 17;
pub const BLOCK_INDEX_ENTRY_SIZE: usize = 44;

#[derive(Debug, Clone)]
pub struct BlockIndexEntry {
    pub min_ts: i64,
    pub max_ts: i64,
    pub byte_offset_start: u64,
    pub byte_offset_end: u64,
    pub point_count: u32,
    pub min_tag_id: u32,
    pub max_tag_id: u32,
}

pub fn encode_udp_batch(points: &[Point]) -> Vec<u8> {
    let mut out = Vec::with_capacity(4 + points.len() * 24);
    out.extend_from_slice(&(points.len() as u32).to_le_bytes());
    for p in points {
        out.extend_from_slice(&p.tag_id.to_le_bytes());
        out.extend_from_slice(&p.ts_epoch.to_le_bytes());
        out.push(p.type_code as u8);
        match (&p.value, p.type_code.fixed_size()) {
            (PointValue::Num(v), Some(1)) => out.push(*v as u8),
            (PointValue::Num(v), Some(2)) => out.extend_from_slice(&(*v as i16).to_le_bytes()),
            (PointValue::Num(v), Some(4)) => {
                if p.type_code == ValueTypeCode::Float32 {
                    out.extend_from_slice(&(*v as f32).to_le_bytes());
                } else {
                    out.extend_from_slice(&(*v as i32).to_le_bytes());
                }
            }
            (PointValue::Num(v), Some(8)) => out.extend_from_slice(&(*v).to_le_bytes()),
            (PointValue::Str(s), None) => {
                out.extend_from_slice(&(s.len() as u32).to_le_bytes());
                out.extend_from_slice(s.as_bytes());
            }
            _ => {}
        }
    }
    out
}

pub fn decode_udp_batch(packet: &[u8]) -> Result<Vec<Point>> {
    if packet.len() < 4 {
        bail!("packet too short");
    }
    let count = u32::from_le_bytes(packet[0..4].try_into().unwrap()) as usize;
    let mut offset = 4usize;
    let mut out = Vec::with_capacity(count);
    for _ in 0..count {
        if offset + 13 > packet.len() {
            bail!("incomplete point header");
        }
        let tag_id = u32::from_le_bytes(packet[offset..offset + 4].try_into().unwrap());
        offset += 4;
        let ts_epoch = i64::from_le_bytes(packet[offset..offset + 8].try_into().unwrap());
        offset += 8;
        let type_code = ValueTypeCode::from_u8(packet[offset]).ok_or_else(|| anyhow::anyhow!("bad type code"))?;
        offset += 1;
        let value = match type_code.fixed_size() {
            Some(1) => PointValue::Num(packet[offset] as i8 as f64),
            Some(2) => {
                let v = i16::from_le_bytes(packet[offset..offset + 2].try_into().unwrap());
                offset += 2;
                PointValue::Num(v as f64)
            }
            Some(4) => {
                if type_code == ValueTypeCode::Float32 {
                    let v = f32::from_le_bytes(packet[offset..offset + 4].try_into().unwrap());
                    offset += 4;
                    PointValue::Num(v as f64)
                } else {
                    let v = i32::from_le_bytes(packet[offset..offset + 4].try_into().unwrap());
                    offset += 4;
                    PointValue::Num(v as f64)
                }
            }
            Some(8) => {
                let v = f64::from_le_bytes(packet[offset..offset + 8].try_into().unwrap());
                offset += 8;
                PointValue::Num(v)
            }
            None => {
                if offset + 4 > packet.len() {
                    bail!("incomplete str len");
                }
                let len = u32::from_le_bytes(packet[offset..offset + 4].try_into().unwrap()) as usize;
                offset += 4;
                if offset + len > packet.len() {
                    bail!("incomplete str payload");
                }
                let s = String::from_utf8_lossy(&packet[offset..offset + len]).to_string();
                offset += len;
                PointValue::Str(s)
            }
            _ => unreachable!(),
        };
        if type_code.fixed_size() == Some(1) {
            offset += 1;
        }
        out.push(Point {
            tag_id,
            ts_epoch,
            type_code,
            value,
        });
    }
    Ok(out)
}

pub fn encode_segment_record(p: &Point) -> Vec<u8> {
    let mut out = Vec::with_capacity(32);
    out.extend_from_slice(&p.tag_id.to_le_bytes());
    out.extend_from_slice(&p.ts_epoch.to_le_bytes());
    out.push(p.type_code as u8);
    match (&p.value, p.type_code.fixed_size()) {
        (PointValue::Num(v), Some(1)) => {
            out.extend_from_slice(&0u32.to_le_bytes());
            out.push(*v as i8 as u8);
        }
        (PointValue::Num(v), Some(2)) => {
            out.extend_from_slice(&0u32.to_le_bytes());
            out.extend_from_slice(&(*v as i16).to_le_bytes());
        }
        (PointValue::Num(v), Some(4)) => {
            out.extend_from_slice(&0u32.to_le_bytes());
            if p.type_code == ValueTypeCode::Float32 {
                out.extend_from_slice(&(*v as f32).to_le_bytes());
            } else {
                out.extend_from_slice(&(*v as i32).to_le_bytes());
            }
        }
        (PointValue::Num(v), Some(8)) => {
            out.extend_from_slice(&0u32.to_le_bytes());
            out.extend_from_slice(&(*v).to_le_bytes());
        }
        (PointValue::Str(s), None) => {
            out.extend_from_slice(&(s.len() as u32).to_le_bytes());
            out.extend_from_slice(s.as_bytes());
        }
        _ => {
            out.extend_from_slice(&0u32.to_le_bytes());
        }
    }
    out
}

pub fn decode_segment_record(buf: &[u8], offset: usize) -> Option<(Point, usize)> {
    if offset + SEGMENT_HEADER_SIZE > buf.len() {
        return None;
    }
    let tag_id = u32::from_le_bytes(buf[offset..offset + 4].try_into().ok()?);
    let ts_epoch = i64::from_le_bytes(buf[offset + 4..offset + 12].try_into().ok()?);
    let type_code = ValueTypeCode::from_u8(buf[offset + 12])?;
    let value_len = u32::from_le_bytes(buf[offset + 13..offset + 17].try_into().ok()?) as usize;
    let payload_len = type_code.fixed_size().unwrap_or(value_len);
    let total = SEGMENT_HEADER_SIZE + payload_len;
    if offset + total > buf.len() {
        return None;
    }
    let payload = &buf[offset + SEGMENT_HEADER_SIZE..offset + total];
    let value = match type_code.fixed_size() {
        Some(1) => PointValue::Num(payload[0] as i8 as f64),
        Some(2) => PointValue::Num(i16::from_le_bytes(payload.try_into().ok()?) as f64),
        Some(4) => {
            if type_code == ValueTypeCode::Float32 {
                PointValue::Num(f32::from_le_bytes(payload.try_into().ok()?) as f64)
            } else {
                PointValue::Num(i32::from_le_bytes(payload.try_into().ok()?) as f64)
            }
        }
        Some(8) => PointValue::Num(f64::from_le_bytes(payload.try_into().ok()?)),
        None => PointValue::Str(String::from_utf8_lossy(payload).to_string()),
        _ => return None,
    };
    Some((
        Point {
            tag_id,
            ts_epoch,
            type_code,
            value,
        },
        total,
    ))
}

pub fn encode_block_index_entry(e: &BlockIndexEntry) -> [u8; BLOCK_INDEX_ENTRY_SIZE] {
    let mut out = [0u8; BLOCK_INDEX_ENTRY_SIZE];
    out[0..8].copy_from_slice(&e.min_ts.to_le_bytes());
    out[8..16].copy_from_slice(&e.max_ts.to_le_bytes());
    out[16..24].copy_from_slice(&e.byte_offset_start.to_le_bytes());
    out[24..32].copy_from_slice(&e.byte_offset_end.to_le_bytes());
    out[32..36].copy_from_slice(&e.point_count.to_le_bytes());
    out[36..40].copy_from_slice(&e.min_tag_id.to_le_bytes());
    out[40..44].copy_from_slice(&e.max_tag_id.to_le_bytes());
    out
}

pub fn decode_block_index(buf: &[u8]) -> Vec<BlockIndexEntry> {
    let mut out = Vec::new();
    let mut off = 0usize;
    while off + BLOCK_INDEX_ENTRY_SIZE <= buf.len() {
        out.push(BlockIndexEntry {
            min_ts: i64::from_le_bytes(buf[off..off + 8].try_into().unwrap()),
            max_ts: i64::from_le_bytes(buf[off + 8..off + 16].try_into().unwrap()),
            byte_offset_start: u64::from_le_bytes(buf[off + 16..off + 24].try_into().unwrap()),
            byte_offset_end: u64::from_le_bytes(buf[off + 24..off + 32].try_into().unwrap()),
            point_count: u32::from_le_bytes(buf[off + 32..off + 36].try_into().unwrap()),
            min_tag_id: u32::from_le_bytes(buf[off + 36..off + 40].try_into().unwrap()),
            max_tag_id: u32::from_le_bytes(buf[off + 40..off + 44].try_into().unwrap()),
        });
        off += BLOCK_INDEX_ENTRY_SIZE;
    }
    out
}
